tiledb/common/thread_pool.cc

```diff
@@ -31,25 +31,41 @@
  */
 
 #include <cassert>
+#include <memory>
 
 #include "tiledb/common/logger.h"
 #include "tiledb/common/thread_pool.h"
 
 namespace tiledb {
 namespace common {
 
-// Define the static ThreadPool member variables.
-std::unordered_map<std::thread::id, ThreadPool*> ThreadPool::tp_index_;
-std::mutex ThreadPool::tp_index_lock_;
-std::unordered_map<std::thread::id, tdb_shared_ptr<ThreadPool::PackagedTask>>
-    ThreadPool::task_index_;
-std::mutex ThreadPool::task_index_lock_;
+template <class T>
+typename ClassSingleton<T>::member_type ClassSingleton<T>::member_factory() {
+  std::lock_guard<std::mutex> lock(lock_central);
```

Comment:
'lock_central' being defined as a static seems to be subject to the same uncertain ordering, tho this time observed in relation to construction of a global ThreadPool entity, relative to when the static lock_central is constructed (in my scenario not yet constructed when the threadpool construction attempt is made), with failure occurring in the construction of the 'lock(lock_central)' at top of 'member_factory()' presumably because the 'lock_central' mutex has not yet been constructed (a nullptr was reported in the error in method trying to lock the mutex.)

Reply:
Apparently MSVC initializes mutexes during dynamic initialization, even though they're declared constexpr and (arguably?) ought to be initialized during constant initialization, that is, before they actually are. I'll get a workaround posted for this.

Reply:
Workaround complete. We define a stat mutex in the factory rather than use a static mutex in the class.